controller/src/lib.rs

```diff
@@ -632,6 +744,253 @@ impl Controller {
+                // module, only which hardware signals it responds to.
+                let need_power_enabled = Self::filter_modules_with(modules, &status, |st| {
+                    !st.contains(Status::POWER_GOOD | Status::ENABLED)
+                })?;
+
+                // Check for any modules which need power applied, but which do
+                // _not_ already have reset asserted. We're going to assert
+                // reset on them now, but emit a warning.
+                let need_reset_deasserted =
+                    Self::filter_modules_with(modules, &status, |st| st.contains(Status::RESET))?;
+                let need_power_but_not_reset = ModuleId {
+                    fpga_id: modules.fpga_id,
+                    ports: need_power_enabled
+                        .ports
+                        .remove(&need_reset_deasserted.ports),
+                };
+                if need_power_but_not_reset.selected_transceiver_count() > 0 {
+                    warn!(
+                        self.log,
+                        "Found modules with power disabled, but reset deasserted. \
+                        It will be asserted before enabling power.";
+                        "need_power_enabled" => ?need_power_enabled,
+                        "need_reset_deasserted" => ?need_reset_deasserted,
+                        "suspicious_modules" => ?need_power_but_not_reset,
+                    );
+                    self.assert_reset(need_power_but_not_reset).await?;
+
+                    // Wait for SFF-8679 `t_reset_init`. It's a bit silly to
+                    // wait 10us here, but we're trying to be careful.
+                    tokio::time::sleep(Duration::from_micros(10)).await;
+                }
+
+                // Enable power and deassert reset for the required modules.
+                //
+                // Note that we are explicitly ensuring above that all modules
+                // which need reset deasserted also need power enabled. (This
+                // cannot apply to modules in high-power mode.) So we can use
+                // the `need_power_enabled` modules for both operations.
+                if need_power_enabled.selected_transceiver_count() > 0 {
+                    self.enable_power(need_power_enabled).await?;
+                }
+                if need_power_enabled.selected_transceiver_count() > 0 {
+                    self.deassert_reset(need_power_enabled).await?;
+
+                    // The SFF-8769 specifies that modules may take up to 2
+                    // seconds after asserting ResetL before they are ready for
+                    // reads. This is `t_reset`.
+                    tokio::time::sleep(Duration::from_secs(2)).await;
+                }
+
+                // Actually write the power mode, the pin and to the memory map.
+                self.set_lp_mode(modules, &current_power_state, mode).await
```

Comment:
To be extra careful we could put this between the `enable_power` and `deassert_reset` above. This ensures that we are asserting `LpMode` _after_ applying power (addressing the voltage problem mentioned above) but _before_ we take the device out of reset. This is definitely belt and suspenders, but figured I'd mention it.

Reply:
Ok, I can do that. It's a bit of reworking, but I already have the methods for it. It looks a bit funky, because we set the hardware / software `LPMode` before and after `deassert_reset`, but I think that only matters when there are modules that we're bringing from off -> low-power. For those going from high -> low or low -> high, those calls to `enable_power` and `assert_reset` don't apply, since those modules aren't in the mask `need_power_enabled`.

Thanks, will update now.

Reply:
Done in 8c46545, let me know what you think @Aaron-Hartwig.

Reply:
Looking at that diff, I think I better understand the "weirdness" around how we have to manage both the hardware and software state involved with low vs high power mode.  https://github.com/oxidecomputer/transceiver-control/commit/8c46545ba16416408f6d3fe498e311d38a962a20 gives what I asked for and definitely is not as "clean" as the previous implementation, but I do think it is a more accurate reflection of what is actually going on.

Reply:
Great, thanks for helping make that accurate! I agree that it's a bit confusing, since we're interspersing the hardware / software mode control between twiddling the power and reset hardware signals. I tried to lay the reasoning out clearly in the block comment, and hopefully neither of us is confused the next time we come back here!